Barry Harley - Mina Frontend Challenge

bmharley opened the pull request.

Humenko commented:
## Overview 
Hey Barry, thought it would be best to give my feedback on here. As a whole, I'm really impressed with the structure and build of this application.

### Things I liked
- Love the `GIVEN / WHEN / THEN` breakdown in readme. Really good way of breaking down the ACs.
- Real nice usage of mock service worker, and implementing more than just mocked tests!
- Clean UI. Aware there is only so much you can take it, but ticks the boxing in terms of layout.

### Things I noticed
- Do you have any standards you stick to around describing your tests? Just a personal opinion, but I've always found it easier to read the terminal log of tests when they are produced as `GIVEN/WHEN/THEN`s. 
- Where applicable I’d always steer away from using `test-id` for react testing. Mainly, this is due to wanting the tests to as closely resemble the user interaction as possible! https://testing-library.com/docs/queries/about/#priority.
- There is some repetition in code where it could be improved, 100% get its a time constrained approach, but would love to know if there would be any improvements you'd make. 

### Extra comments
- Have you built any nextjs applications at all? Would love your opinions on it too if you have.
- If you are new to typescript, I found a really useful/(king of fun) repo when I was learning that you could have a read of to understand more of typings! https://github.com/type-challenges/type-challenges
- No opinions here on your implementation, more of an fyi, and thought it would be useful to let you know, that we use css-in-js, more particularly https://styled-components.com/. Incase you had any opinions, or wanted to have a read at this different implementation of styling.

Just hit me up here or on email if you want to discuss more! But great work!

bmharley commented:
> ## Overview
> Hey Barry, thought it would be best to give my feedback on here. As a whole, I'm really impressed with the structure and build of this application.
> 
> ### Things I liked
> * Love the `GIVEN / WHEN / THEN` breakdown in readme. Really good way of breaking down the ACs.
> * Real nice usage of mock service worker, and implementing more than just mocked tests!
> * Clean UI. Aware there is only so much you can take it, but ticks the boxing in terms of layout.
> 
> ### Things I noticed
> * Do you have any standards you stick to around describing your tests? Just a personal opinion, but I've always found it easier to read the terminal log of tests when they are produced as `GIVEN/WHEN/THEN`s.
> * Where applicable I’d always steer away from using `test-id` for react testing. Mainly, this is due to wanting the tests to as closely resemble the user interaction as possible! https://testing-library.com/docs/queries/about/#priority.
> * There is some repetition in code where it could be improved, 100% get its a time constrained approach, but would love to know if there would be any improvements you'd make.
> 
> ### Extra comments
> * Have you built any nextjs applications at all? Would love your opinions on it too if you have.
> * If you are new to typescript, I found a really useful/(king of fun) repo when I was learning that you could have a read of to understand more of typings! https://github.com/type-challenges/type-challenges
> * No opinions here on your implementation, more of an fyi, and thought it would be useful to let you know, that we use css-in-js, more particularly https://styled-components.com/. Incase you had any opinions, or wanted to have a read at this different implementation of styling.
> 
> Just hit me up here or on email if you want to discuss more! But great work!

Hey @Humenko , thanks again for taking to the time to provide some constructive feedback and good to hear you were happy with the solution overall .

Below, I've replied to each of your feedback comments.

### Things I noticed

- **GIVEN/WHEN/THEN to describe tests:** typically I tend to squash the test description into a single sentence, which can become verbose and hard to read (and also lacks consistency as you pointed out). So I'm looking to improve in this area and love the idea of taking a BDD approach. I've added a **new commit** where I've applied this to the CommitsList component as a practise exercise, any feedback would be appreciated. Will definitely be using this more in the future. A few other thoughts below.
-- I've been careful to not nest too deeply here so it doesn't become hard to read, sticking to 3 levels (test suite name -> GIVEN -> WHEN -> THEN). I'd imagine its easy to end up in a tangle here.
-- I went back and fourth a few times on whether 'GIVEN' is actually adding any value in some cases (or whether it's just making it more verbose) e.g. when it's just rendering a component with no parameters possibly better to skip straight to WHEN here.
-- Would be interested to get your opinion on whether you prefer to split tests up into granular cases. Whereby you test one thing only, or whether you're more into the idea of grouping related expects together. For example, for my CommitsList component I have a bunch tests to check that page two is shown when you click 'next'. I've split up the various expects here into separate tests cases e.g. commits for page 2 are shown, previous button enabled, next button disabled, page counter is updated. Alternatively I could have placed all these expects under one test e.g. 'THEN page 2 is shown'
- **Use of testId:** I was aware that this is (generally) bad practice. There were a few instances here where I made use of it purely from a time saving perspective. In my new commit I've **purged all uses of testId**.
- **Code repetition:** Definitely room for improvement here - particularly around the reducer (with exception of commits the rest of the switch cases suffer from duplication). Over the coming days I'll set some time aside to have a think about this in more detail, how to make it more DRY, and come back to you with some examples on how I'd approach refactoring.

### Extra comments

- **NextJS** - on my Trello 'things to learn' board - I'll be honest it's not something I have any experience with so I'm approaching it from scratch but looking forward to getting stuck in.

- **Typescript** - I'm relatively new to (it's currently top of my list on personal development), slowly getting there with it, some of it has been tricky to wrap my head around particularly generics and passing type parameters but I'm getting there. This tech test served as a good learning exercise. Will be sure to check out **type-challenges**; which looks a fun way to learn, thanks for sharing.

- **css-in-js** - will be sure to take this for a test drive too... certainly sound an interesting concepts of removing the mapping between components and styles.
